Update InputDatePickerFormField callbacks (onDateSubmitted, onDateSaved)

RamonFarizel opened the pull request:
Fixed the issue where `InputDatePickerFormField` didn't call `onDateSubmitted` and `onDateSaved` with `null` when `acceptEmptyDate: true` and the field was cleared.

I'm not quite sure if this is the recommended approach, so I implemented it based on the current discussion in the issue.
In addition to it, it might be a breaking change. 



Fixes: https://github.com/flutter/flutter/issues/179142


## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

RamonFarizel committed:
allow onDateSubmitted and onDateSaved to return null

RamonFarizel committed:
improve readability

RamonFarizel committed:
fix dart analyzer issues

RamonFarizel committed:
Merge branch 'master' into rf/date_picker_empty_field

MitchellGoodwin reviewed (commented):
You are correct in that this is a breaking change, as the types would need to be updated for anyone providing these callbacks.

https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes

We need to leave the current types for onDateSaved and onDateSubmitted alone.

Three approaches we could take

1. Add new versions of those two properties under new names and deprecate the old ones.

2. Add an `onDateCleared` property to call when the date is made null in the input field to report that outside of the field.

3. Add an `onDateChanged` of type `ValueChanged<DateTime?>?`. This would be called on empty input field, as well as on both `onDateSubmitted` and `onDateSaved`.

In my opinion, option 1 would be a little clunky. Option 2 would work, but would be some API bloat for just watching for when the date is cleared. Option 3 would be my preference because it would work for this case, but also would be one place for handling what I suspect is the main usecase developers have for these APIs: watching for changes to the date in the internal state of the input field.

I'd like @dkwingsmt's opinion on which approach to take.

dkwingsmt commented:
Sorry for the late reply! I also agree with option 3. Thank you for working on this.

dkwingsmt commented:
Hi @RamonFarizel Just want to make sure if you're good to proceed, or are you waiting for anything else. It's ok if you're busy right now. I want to make sure I didn't miss anything. Thank you!

RamonFarizel committed:
Merge branch 'master' into rf/date_picker_empty_field

RamonFarizel committed:
Remove first approach

RamonFarizel committed:
Add onDateChanged callback to InputDatePickerFormField

- Introduced onDateChanged callback to handle date value changes, including null when the field is cleared.
- Updated tests to verify the behavior of onDateChanged for various scenarios, including valid date submissions and empty field handling.

RamonFarizel committed:
Merge branch 'master' into rf/date_picker_empty_field

RamonFarizel committed:
fix format

RamonFarizel committed:
Merge branch 'master' into rf/date_picker_empty_field

RamonFarizel committed:
Merge branch 'master' into rf/date_picker_empty_field

dkwingsmt commented:
Thanks for your contribution! @RamonFarizel  Since I haven't seen active progress lately, I'm going to reluctantly close this pull request for now, but if you get time to address the comments, we'd be more than happy to take a look at a new patch. If you do send a new patch, it would be very useful to reference this one in the description in order to help reviewers see the previous context.
